sympy/polys/polytools.py

```diff
@@ -3773,6 +3774,178 @@ def nth_power_roots_poly(f, n):
 
         return r.replace(t, x)
 
+    def root_bounds(f):
+        """
+        Compute lower and upper bounds on the magnitude of the nonzero roots of
+        this polynomial.
+
+        Examples
+        ========
+
+        >>> from sympy import Poly
+        >>> from sympy.abc import x
+        >>> Poly(x**2 - 2).root_bounds()
+        (2/3, 3)
+        >>> (x * Poly(x**2 - 2)).root_bounds()
+        (2/3, 3)
+
+        Returns
+        =======
+
+        Pair ``(lb, ub)`` of :py:class:`~.Expr`
+            Gives the lower bound and upper bound. These bounds will be
+            rational if the domain of the polynomial is :ref:`ZZ` or :ref:`QQ`.
+
+        Raises
+        ======
+
+        DomainError
+            If the domain of the polynomial is not a subring of the complex
+            numbers.
+        MultivariatePolynomialError
+            If the polynomial is not univariate.
+        PolynomialError
+            If the polynomial is a monomial, since in such a case it has no
+            nonzero roots.
+
+        Notes
+        =====
+
+        The domain of the polynomial must be a subring of the complex numbers.
+        We use the `Cauchy bound`_.
+
+        .. _Cauchy bound: https://en.wikipedia.org/wiki/Geometrical_properties_of_polynomial_roots#Lagrange's_and_Cauchy's_bounds
+
+        """
+        dom = f.get_domain()
+        if dom.is_FiniteField or not dom.is_Numerical:
+            raise DomainError("Cannot compute root bounds over %s" % dom)
+        if f.is_multivariate:
+            raise MultivariatePolynomialError(
+                "Must be a univariate polynomial")
+        if f.is_monomial:
+            raise PolynomialError("Monomial has no nonzero roots")
+
+        coeffs = f.all_coeffs()
+        while coeffs[-1] == 0:
+            coeffs.pop()
+        lc, tc = coeffs[0], coeffs[-1]
+        ub = 1 + max(abs(n / lc) for n in coeffs[1:])
+        lb_inv = 1 + max(abs(n / tc) for n in coeffs[:-1])
```

Comment:
It does seem somewhat backwards to use Expr here. Can we not convert to a field and compute this with domain elements e.g.:
```python
In [24]: p = Poly(x**2 + 2)

In [25]: p = p.to_field()

In [26]: coeffs = p.rep.rep

In [27]: result = 1 + max(abs(n/coeffs[0]) for n in coeffs[1:])

In [28]: result
Out[28]: mpq(3,1)

In [29]: K.to_sympy(result)
Out[29]: 3
```

Reply:
Actually running into an interesting issue here, depending on the domain of the `Poly`, since `abs()` is not supported for domain elements of types `ANP` or `GuassianElement`.

What do you suppose is the right way to handle this? Should these domain types define an `__abs__()` method?

For example, I tried giving `GaussianRational` this method:

```python
    def __abs__(self):
        from sympy.polys.domains.complexfield import CC
        z = CC.from_GaussianRationalField(self, self.base)
        return abs(z)
```

which does solve the immediate problem. But unfortunately importing `CC` at the top of the module causes a cyclic import, and importing it here seems a bit ugly. Is there a better way?

Reply:
I still think that we should try to avoid introducing square roots if it's possible.

Maybe it would be better to give a domain method like `K.abs2` which can compute `abs(z)**2` since I think it should always be possible to compute that within the same ring:
```python
def abs2(K, z):
    return z.x**2 + z.y**2
```
For ANP it could be done if there was a way to compute complex conjugates...

Reply:
Yeah, if it's important to stay in the Poly's domain, then I think there will be trouble with `ANP` (e.g. the fields generated by the cube roots of 2 -- the universal counterexample!)....

Also if `Poly.root_bounds()` returns squared bounds, then the `Poly.root_separation_...` method will return the 4th power of the bound, and I feel like it's getting out of hand...

I'm thinking I'd really like to retreat a bit here. How about if `Poly.root_bounds()` only supports polys over `QQ` and `ZZ`. That's all that was originally called for. I was trying to support all domains contained in `CC` because it seemed easy, but it no longer does. If someone needs that at some point, let them open an issue at that time.